### PR TITLE
Handle index pattern selector on new discover

### LINKED ISSUE #6434: [Remove discover] Change index pattern selector mechanism
## Description

Make the search bar use the index selected in the index pattern selector. 

Screenshot 2024-02-20 at 12 39 23

## Preconditions

The user must add a custom index pattern in the `App Settings > Index pattern` section 

Screenshot 2024-02-20 at 12 40 24

That will show the index pattern selector in the header.

## Uses cases

- When the user selects the index pattern using the selector. The discover (events tab) must use the selected index pattern.

## Discover

- When the discover component receives a default index pattern like prop. The discover must use it (The component has to receive the default index pattern with an IndexPattern type)
- When the discover component does not receive a default index pattern like prop the component must get the current index pattern from the AppState.

### before

```ts
type WazuhDiscoverProps {
  indexPatternName: string
  tableColumns: tDataGridColumn
}
```

### after

```ts
type WazuhDiscoverProps {
  indexPatternName: IndexPattern
  tableColumns: tDataGridColumn
}
```

## Tasks

- [x] Research about index pattern selector feature
- [x] Create a generic index pattern selector component

### PR DESCRIPTION
**Pull request by Machi3mfl**
### Description
This PR, adds the feature to the discover component to use the index pattern selected in the index pattern selector positioned on the header.

Closes #6434 

## Preconditions

- Create an index pattern that matches with the wazuh-alerts-* pattern (wazuh-a*)
- Create an index pattern that doesn't matches with the wazuh-alerts-* pattern (wazuh-custom-*)

Use the App Settings > Index pattern section to create both index patterns

Screenshot 2024-03-11 at 17 47 29

## Index pattern selector

| Test | Chrome | Firefox | Safari |
| --- |  --- | --- | --- |
| Check the index pattern list in the select | :black_circle: | :black_circle: | :black_circle: |
| Check the index pattern selected by default in the select match with the stored in browser cookies | :black_circle: | :black_circle: | :black_circle: |
| Change index pattern and check if was stored in browser cookies | :black_circle: | :black_circle: | :black_circle: |
| Check toast error when is selected an index pattern without indices (wazuh-custom-* index) | :black_circle: | :black_circle: | :black_circle: |

<details>
<summary>
Check the index pattern list in the select
</summary>

Compare the index patterns shown in the select with the index pattern on the Dashboard Management (only must appear the alerts indexes, exclude monitoring and statistics)

Screenshot 2024-03-11 at 17 48 51

image

</details>

<details>
<summary>
Check the index pattern selected by default in the select match with the stored in browser cookies
</summary>

image

</details>

<details>
<summary>
Change index pattern using the select and check if is stored in browser cookies
</summary>

When the current tab is not the dashboard tab, the app reloads the window

https://github.com/wazuh/wazuh-dashboard-plugins/assets/6089438/7cbd4bb8-0d4f-4e55-b617-8b44647114f0

</details>

<details>
<summary>
Check toast error when is selected an index pattern without indices (wazuh-custom-* index)</summary>

The cookie must not change and preserve the previous index

https://github.com/wazuh/wazuh-dashboard-plugins/assets/6089438/df614607-ae87-4a17-93db-3e9baee45bf7

</details>

### Check List
- [x] All tests pass
  - [x] `yarn test:jest`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: plugins/main/public/components/common/data-source/pattern/pattern-data-source-repository.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
+import { tDataSource } from "../data-source";
+import { DataSourceRepository } from "../data-source-repository";
+import { PatternDataSource } from "./pattern-data-source";
+import { GenericRequest } from '../../../../react-services/generic-request';
+import { AppState } from '../../../../react-services';
+
+/*
```
Comment (yenienserrano) [inline plugins/main/public/components/common/data-source/pattern/pattern-data-source-repository.ts:7]: Is this necessary or can it be deleted?

**Comment by yenienserrano** [comment]
| Test | Chrome | Firefox | Safari |
| --- |  --- | --- | --- |
| Check the index pattern list in the select | :black_circle: | :black_circle: | :black_circle: |
| Check the index pattern selected by default in the select match with the stored in browser cookies | :black_circle: | :black_circle: | :black_circle: |
| Change index pattern and check if was stored in browser cookies | :black_circle: | :black_circle: | :black_circle: |
| Check toast error when is selected an index pattern without indices (wazuh-custom-* index) | :black_circle: | :black_circle: | :black_circle: |

<details>
<summary>
Check the index pattern list in the select
</summary>

Compare the index patterns shown in the select with the index pattern on the Dashboard Management (only must appear the alerts indexes, exclude monitoring and statistics)


Screenshot 2024-03-11 at 17 48 51

image

</details>

<details>
<summary>
Check the index pattern selected by default in the select match with the stored in browser cookies
</summary>

image


</details>



<details>
<summary>
Change index pattern using the select and check if is stored in browser cookies
</summary>

When the current tab is not the dashboard tab, the app reloads the window

https://github.com/wazuh/wazuh-dashboard-plugins/assets/6089438/7cbd4bb8-0d4f-4e55-b617-8b44647114f0

</details>


<details>
<summary>
Check toast error when is selected an index pattern without indices (wazuh-custom-* index)</summary>

The cookie must not change and preserve the previous index

https://github.com/wazuh/wazuh-dashboard-plugins/assets/6089438/df614607-ae87-4a17-93db-3e9baee45bf7

</details>

**Comment by yenienserrano** [comment]
When you have more than one index pattern and you refresh the page, the selector does not appear until you navigate to another application.

https://github.com/wazuh/wazuh-dashboard-plugins/assets/63758389/0cbdf572-a238-42d0-83a0-ed3fb18441c9

**Comment by lucianogorza** [comment]
## CR 🟢
## Test 🟢

| Test | Chrome | Firefox | Safari |
| --- |  --- | --- | --- |
| Check the index pattern list in the select | 🟢 | :black_circle: | :black_circle: |
| Check the index pattern selected by default in the select match with the stored in browser cookies | 🟢 | :black_circle: | :black_circle: |
| Change index pattern and check if was stored in browser cookies | 🟢 | :black_circle: | :black_circle: |
| Check toast error when is selected an index pattern without indices (wazuh-custom-* index) | 🟢 | :black_circle: | :black_circle: |

<details>
<summary>
Check the index pattern list in the select
</summary>

Compare the index patterns shown in the select with the index pattern on the Dashboard Management (only must appear the alerts indexes, exclude monitoring, statistics and vulnerabilities)

image
image

</details>

<details>
<summary>
Check the index pattern selected by default in the select match with the stored in browser cookies
</summary>

image

</details>

<details>
<summary>
Change index pattern using the select and check if is stored in browser cookies
</summary>

[change-index-pattern.webm](https://github.com/wazuh/wazuh-dashboard-plugins/assets/103193307/450c5631-bb24-4edb-8a03-87fa062fff01)

</details>


<details>
<summary>
Check toast error when is selected an index pattern without indices (wazuh-custom-* index)</summary>

The cookie must not change and preserve the previous index

[change-index-pattern-error.webm](https://github.com/wazuh/wazuh-dashboard-plugins/assets/103193307/72213520-26e1-4fe4-9347-23e903bb55f8)

</details>

**Review by lucianogorza — APPROVED** [review]
LGTM

File: plugins/main/public/components/common/data-source/pattern/pattern-data-source-repository.ts (hunk 1 above)
Comment (Machi3mfl) [reply plugins/main/public/components/common/data-source/pattern/pattern-data-source-repository.ts]: Removed in 6583f4da21c0fc54d8fd3f0cbe98e79e52de83d9

**Pull request merged**